A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+
+Following code snippets show how to enable authorization in gRPC servers in 
+different languages.
+
+#### C++
+
+Static Initialization
+
+```C++
+std::shared_ptr<AuthorizationPolicyProviderInterface> provider = 
+	StaticDataAuthorizationPolicyProvider::Create(authz_policy);
+ServerBuilder builder;
+builder.SetAuthorizationPolicyProvider(provider);
+std::unique_ptr<grpc::Server> server = builder.BuildAndStart();
+```
+
+Dynamic file reloading
+
+```C++
+std::shared_ptr<AuthorizationPolicyProviderInterface> provider = 
+	FileWatcherAuthorizationPolicyProvider::Create(authz_policy_path, /*refresh_interval_sec=*/3600);
+ServerBuilder builder;
+builder.SetAuthorizationPolicyProvider(provider);
+std::unique_ptr<grpc::Server> server = builder.BuildAndStart();
+```
+
+#### Go
+
+Static initialization
+
+```Go
+creds := credentials.NewServerTLSFromFile(certFile, keyFile)
+i, err := authz.NewStatic(authzPolicy)
+serverOpts := []grpc.ServerOption{
+  grpc.Creds(creds),
+  grpc.ChainUnaryInterceptor(i.UnaryInterceptor),
+  grpc.ChainStreamInterceptor(i.StreamInterceptor),
+}
+s := grpc.NewServer(serverOpts...)
+```
+
+Dynamic File Reloading
+
+```Go
+creds := credentials.NewServerTLSFromFile(certFile, keyFile)
+i, err := authz.NewFileWatcher(authzPolicyFile, 1 * time.Hour)
+serverOpts := []grpc.ServerOption{
+  grpc.Creds(creds),
+  grpc.ChainUnaryInterceptor(i.UnaryInterceptor),
+  grpc.ChainStreamInterceptor(i.StreamInterceptor),
+}
+s := grpc.NewServer(serverOpts...)
+// In the end, free up the resources used for policy refresh.
+defer i.Close()
+```
+
+#### Java
+
+Static initialization
+
+```Java
+AuthorizationServerInterceptor authzServerInterceptor;
+try {
+  authzServerInterceptor = AuthorizationServerInterceptor.create(authzPolicy);
+} catch (IllegalArgumentException iae) {
+  // ...
+}
+Server server =
+    ServerBuilder.forPort(port)
+        .addService(service)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .intercept(authzServerInterceptor);
+        .build()
+        .start();
+```
+
+Dynamic file reloading
+
+```Java
+ScheduledExecutorService scheduledExecutor =
+    Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder()
+            .setNameFormat(...)
+            .setDaemon(true)
+            .build());
+FileAuthorizationServerInterceptor authzServerInterceptor;
+try {
+  authzServerInterceptor = FileAuthorizationServerInterceptor.create(authzPolicyFile);
+} catch (IllegalArgumentException iae) {
+  // ...
+} catch (IOException ioe) {
+  // ...
+}
+Closeable closeable = authzServerInterceptor.scheduleRefreshes(
+    1, TimeUnit.HOURS, scheduledExecutor);
+Server server =
+    ServerBuilder.forPort(port)
+        .addService(service)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .intercept(authzServerInterceptor)
+        .build()
+        .start();
+// In the end, free up the resources used for policy refresh.
+closeable.close();
+```
```

Comment:
Can this also be simplified to eliminate a lot of boilerplate?
```
FileAuthorizationPolicyProvider fileAuthzProvider 
     = FileAuthorizationPolicyProvider.create(
               authzPolicyFile, scheduledExecutor, 1, TimeUnit.HOURS);

Server server =    ServerBuilder.forPort(port)
        .addService(service)
        .useTransportSecurity(certChainFile, privateKeyFile)
        .authzPolicyProvider(fileAuthzProvider);
        .build()
        .start();
```

1) The same `authzPolicyProvider()` method is used and it does the right setup based on whether received provider is static or dynamicFile based
2) The closeable is also internally managed by the server where is will call close on the internally created Closeable

The only downside I see is this cannot be shared among multiple Servers.

Also consider using `io.grpc.Grpc.newServerBuilderForPort(int port, ServerCredentials creds)` in the example

Reply:
Based on the email discussion, we won't be updating the current design.